bliss/models/fnp.py

```diff
@@ -0,0 +1,475 @@
+
+    The FNP is made up of a few different modules. These are all
+    application-specific, so they are passed as arguments to the
+    class constructor
+    * cov_vencoder This is a module which takes input tensor X
+    and outputs a probability distribution from which U is sampled.
+    * dep_graph: An object of type DepGraph which samples graphs G and A.
+    Requires methods .sample_G(uR) and .sample_A(uM, uR)
+    * trans_cond_y: A module which takes labels Y and returns a flattened
+    representation Y_encoded
+    * rep_encoder: This is a module which takes (u, uR, XR, yR_encoded) as input
+    and outputs a flattened representation. (see RepEncoder)
+    * pooler: This is a module which takes the output of rep_encoder and the dependency matrix
+    and returns a sampler for the latent variables Z. (see AveragePooler, SetPooler)
+    * prop_vencoder: This is a module which samples a representation
+    for an object given X and yR_encoded (only used in training, not prediction)
+    * label_vdecoder: This is a module which probabilistically
+    decodes the pooled representation into the output
+    * fb_z: If non-zero, the amount of "free-bits" regularization to apply while training. This
+    encourages learning a better representation and can avoid a local minimum in the prior.
+    """
+
+    def __init__(
+        self,
+        cov_vencoder,
+        dep_graph,
+        trans_cond_y,
+        rep_encoder,
+        pooler,
+        prop_vencoder,
+        label_vdecoder,
+        fb_z=0.0,
+    ):
+        super().__init__()
+        ## Learned Submodules
+        self.cov_vencoder = cov_vencoder
+        self.dep_graph = dep_graph
+        self.trans_cond_y = trans_cond_y
+        self.rep_encoder = rep_encoder
+        self.pooler = pooler
+        self.prop_vencoder = prop_vencoder
+        self.label_vdecoder = label_vdecoder
+
+        ## Initialize free-bits regularization
+        self.fb_z = fb_z
+        self.register_buffer("lambda_z", torch.tensor(1e-8))
+
+    def encode(self, XR, yR, XM, G_in=None, A_in=None, mode="infer"):
+        """
+        This method runs the FNP up to the point the distributions for the latent
+        variables are calculated. This is the shared procedure for both model inference
+        and prediction.
+        """
+        n_ref = XR.size(0)
+        X_all = torch.cat([XR, XM], dim=0)
+
+        ## Sample covariate representation U
+        pu = self.cov_vencoder(X_all)
+        u = pu.rsample()
+        uR = u[:n_ref]
+        uM = u[n_ref:]
+        assert torch.isnan(u).sum() == 0
+
+        ## Sample the dependency matrices
+        ## If we are training ("infer"), the entire dependency
+        ## graph (A and G) is generated.
+        ## If we are predicting ("predict"), only the dependent
+        ## graph (A) is used.
+        if A_in is None:
+            A = self.dep_graph.sample_A(uM, uR)
+        else:
+            A = A_in
+        if mode == "infer":
```

Comment:
We might need to change this later to work with pytorch lightning (which has separate training, validation, and testing steps). For now I think you can get rid of this additional flag using the `.training` attribute which should be a part of `nn.Module` by default.